db/write_thread.cc

```diff
@@ -428,51 +428,79 @@ size_t WriteThread::EnterAsBatchGroupLeader(Writer* leader,
   // (newest_writer) is inclusive. Iteration goes from old to new.
   Writer* w = leader;
+  // write_group end
+  Writer* we = leader;
+  // declare r_list
+  Writer* rb = nullptr;
+  Writer* re = nullptr;
+
   while (w != newest_writer) {
     w = w->link_newer;
 
-    if (w->sync && !leader->sync) {
-      // Do not include a sync write into a batch handled by a non-sync write.
-      break;
-    }
-
-    if (w->no_slowdown != leader->no_slowdown) {
-      // Do not mix writes that are ok with delays with the ones that
-      // request fail on delays.
-      break;
-    }
-
-    if (!w->disable_wal && leader->disable_wal) {
-      // Do not include a write that needs WAL into a batch that has
-      // WAL disabled.
-      break;
-    }
-
-    if (w->batch == nullptr) {
-      // Do not include those writes with nullptr batch. Those are not writes,
-      // those are something else. They want to be alone
-      break;
-    }
-
-    if (w->callback != nullptr && !w->callback->AllowWriteBatching()) {
-      // dont batch writes that don't want to be batched
-      break;
+    if ((w->sync && !leader->sync) ||
+        // Do not include a sync write into a batch handled by a non-sync write.
+        (w->no_slowdown != leader->no_slowdown) ||
+        // Do not mix writes that are ok with delays with the ones that request
+        // fail on delays.
+        (w->disable_wal != leader->disable_wal) ||
+        // Do not mix writes that are different request for WAL disabled.
+        (w->batch == nullptr) ||
+        // Do not include those writes with nullptr batch. Those are not writes
+        // those are something else. They want to be alone
+        (w->callback != nullptr && !w->callback->AllowWriteBatching()) ||
+        // dont batch writes that don't want to be batched
+        (size + WriteBatchInternal::ByteSize(w->batch) > max_size)
+        // Do not make batch too big
+    ) {
+      // remove from list
+      w->link_older->link_newer = w->link_newer;
+      if (w->link_newer != nullptr) {
+        w->link_newer->link_older = w->link_older;
+      }
+      // insert into r_list
+      if (re == nullptr) {
+        rb = re = w;
+        w->link_older = nullptr;
+      } else {
+        w->link_older = re;
+        re->link_newer = w;
+        re = w;
+      }
+    } else {
+      // grow up
+      we = w;
+      w->write_group = write_group;
+      size += WriteBatchInternal::ByteSize(w->batch);
+      write_group->last_writer = w;
+      write_group->size++;
     }
-
-    auto batch_size = WriteBatchInternal::ByteSize(w->batch);
-    if (size + batch_size > max_size) {
-      // Do not make batch too big
-      break;
+  }
+  // append r_list after write_group end
+  if (rb != nullptr) {
+    rb->link_older = we;
+    re->link_newer = nullptr;
+    we->link_newer = rb;
+    if (!newest_writer_.compare_exchange_weak(w, re)) {
+      while (w->link_older != newest_writer) {
```

Comment:
Can we use `newest_writer->link_newer->link_order = re` to avoid this while loop?

Reply:
no, `newest_writer->link_newer` will be nullptr, seek code of `LinkOne` & `CreateMissingNewerLinks`